File: SortVision/public/code/bubble/javascript/bubbleSort.js
/*
Bubble Sort Algorithm in JavaScript 

What this file includes :-

-Implemented Bubble Sort Algorithm
-Comments to make code clear
-Time and Space Complexity Analysis
-Example usage
-Test cases
-Performance Optimization Notes

Bubble Sort Algorithm - Compares adjacent elements over n-1 iterations and swaps the numbers to increasing order.

*/

//Implementation

function bubbleSort(arr) {
  const n = arr.length;

  //Iterations
  for(let i = 0; i < n-1; i++) {
    
    //Optimization
    let isSwapped = false;

    //Comparing the adjacent elements and swaping them
    for(let j = 0; j < n-i-1; j++) {

      //swap
      if(arr[j] > arr[j+1]) {
        [arr[j], arr[j+1]] = [arr[j+1], arr[j]];
        isSwapped = true;
      }
    }
    //Optimization - if array is already sorted
    if(!isSwapped) break;
  }

  return arr;
}

/*
Time Complexity

-Iterations - O(n)
-Comparisons - O(n)
-Overall Time complexity as they are loops 
O(n*n) = O(n^2)

Space Complexity

-In place substitutions - O(1)
*/

//Example Usage 

const example = [4, 1, 3, 2, 10, 7];
console.log("Original: ", example.slice());
console.log("Sorted: ", bubbleSort(example.slice()));

//Test Cases

function testCases() {
  //Assigning test cases
  const tests = [
    {input: [], expected: []},
    {input: [1], expected: [1]},
    {input: [2, 1], expected: [1, 2]},
    {input: [2, 2, 2], expected: [2, 2, 2]},
    {input: [10, 3, 1, 2, 4, 11, 7, 6], expected: [1, 2, 3, 4, 6, 7, 10, 11]},
    {input: [1, 2, 1, 2, 1, 2, 1], expected: [1, 1, 1, 1, 2, 2, 2]}
  ];
  
  //Checking tests
  tests.forEach(({ input, expected }, index) => {
    const result = bubbleSort(input.slice());
    const passed = JSON.stringify(result) === JSON.stringify(expected);

    //Printing Checks of test cases
    console.log(`Test ${index+1}:`, passed ? 'Passed!' : `Failed! got ${result} as result`);
  });
}

testCases();

/* Performance Optimization Notes :
- Stop early is array is already sorted using boolean isSwap variable. 
- This is best case as this is O(n) time complexity.
- After every pass, the largest unsorted element is at the correct position.
- No need to check the last i elements again—they are already sorted.
*/

File: SortVision/public/code/bubble/python/bubbleSort.py
"""
Bubble Sort Implementation in Python
Description: Sorts a list of integers using the bubble sort algorithm.
"""
from typing import List

def bubble_sort(arr: List[int]) -> List[int]:
    """
    Sorts a list using the Bubble Sort algorithm.

    Args:
        arr: List of integers to sort

    Returns:
        List[int]: Sorted list

    Time Complexity: O(n^2) in worst and average cases, O(n) in best case (already sorted)
    Space Complexity: O(1) as it's an in-place sorting algorithm
    """
    n = len(arr)

    # If array is empty or has one element, it's already sorted
    if n <= 1:
        return arr

    # Traverse through all array elements
    for i in range(n):
        swapped = False  # To optimize: stop if no swaps in inner loop

        # Last i elements are already in place
        for j in range(0, n - i - 1):
            # Swap if the element found is greater than the next element
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True

        # If no elements were swapped, array is sorted
        if not swapped:
            break

    return arr
